sdks/python/apache_beam/examples/per_entity_training.py

```diff
@@ -0,0 +1,145 @@
+about 15 different attributes like salary >=50k, education level,
+native country, age, occupation and others. The pipeline does some filtering
+by selecting certain education level, discarding missing values and empty rows.
+The pipeline then groups the rows based on education level and
+trains Decision Trees for each group and finally saves them.
+"""
+
+import argparse
+import logging
+import os
+
+from joblib import dump
+import pandas as pd
+from sklearn.compose import ColumnTransformer
+from sklearn.pipeline import Pipeline
+from sklearn.preprocessing import LabelEncoder
+from sklearn.preprocessing import MinMaxScaler
+from sklearn.preprocessing import OneHotEncoder
+from sklearn.tree import DecisionTreeClassifier
+
+import apache_beam as beam
+from apache_beam.options.pipeline_options import PipelineOptions
+from apache_beam.options.pipeline_options import SetupOptions
+
+
+class CreateKey(beam.DoFn):
+  def process(self, element, *args, **kwargs):
+    # 3rd column of the dataset is Education
+    idx = 3
+    key = element.pop(idx)
+    yield (key, element)
+
+
+def custom_filter(element):
+  """Discard data point if contains ?, and
+  doesn't have all features, and
+  doesn't have Bachelors, Masters or a Doctorate Degree"""
+  return len(element) == 15 and '?' not in element \
+      and ' Bachelors' in element or ' Masters' in element \
+      or ' Doctorate' in element
+
+
+class PrepareDataforTraining(beam.DoFn):
+  """Preprocess data in a format suitable for training."""
+  def process(self, element, *args, **kwargs):
+    key, values = element
+    #Convert to dataframe
+    df = pd.DataFrame(values)
+    last_ix = len(df.columns) - 1
+    X, y = df.drop(last_ix, axis=1), df[last_ix]
+    # select categorical and numerical features
+    cat_ix = X.select_dtypes(include=['object', 'bool']).columns
+    num_ix = X.select_dtypes(include=['int64', 'float64']).columns
+    # label encode the target variable to have the classes 0 and 1
+    y = LabelEncoder().fit_transform(y)
+    yield (X, y, cat_ix, num_ix, key)
+
+
+class TrainModel(beam.DoFn):
+  """Takes preprocessed data as input,
+  transforms categorical columns using OneHotEncoder,
+  normalizes numerical columns and then
+  fits a decision tree classifier.
+  """
+  def process(self, element, *args, **kwargs):
+    X, y, cat_ix, num_ix, key = element
+    steps = [('c', OneHotEncoder(handle_unknown='ignore'), cat_ix),
+             ('n', MinMaxScaler(), num_ix)]
+    # one hot encode categorical, normalize numerical
+    ct = ColumnTransformer(steps)
+    # wrap the model in a pipeline
+    pipeline = Pipeline(steps=[('t', ct), ('m', DecisionTreeClassifier())])
+    pipeline.fit(X, y)
+    yield (key, pipeline)
+
+
+class SaveModel(beam.DoFn):
+  """Saves the trained model to specified location."""
+  def process(self, element, path, *args, **kwargs):
+    key, trained_model = element
+    dump(trained_model, os.path.join(path, f"{key}_model.joblib"))
```

Comment:
Instead of using dump with a fixed path, could we please use [fileio with dynamic destinations](https://beam.apache.org/releases/pydoc/2.15.0/apache_beam.io.fileio.html#dynamic-destinations)? That way its easy to write to gcs or other known filesystems.

It would also be good to make that path a configurable optional known_arg